src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Document.cs

```diff
@@ -42,58 +43,102 @@ private void ScanTagInDocumentContext()
-                        Optional(HtmlSymbolType.Text);
-                        TagContent(); // Parse the tag, don't care about the content
-                        Optional(HtmlSymbolType.ForwardSlash);
-                        Optional(HtmlSymbolType.CloseAngle);
-
-                        if (scriptTag)
-                        {
-                            Output(SpanKind.Markup);
-                            tagBlock.Dispose();
-
-                            SkipToEndScriptAndParseCode();
-                            return;
-                        }
-                    }
-                    else
+                AcceptAndMoveNext(); // Accept '<'
+
+                if (!At(HtmlSymbolType.ForwardSlash))
+                {
+                    OptionalBangEscape();
+
+                    // Parsing a start tag
+                    var scriptTag = At(HtmlSymbolType.Text) &&
+                                    string.Equals(CurrentSymbol.Content, "script", StringComparison.OrdinalIgnoreCase);
+                    Optional(HtmlSymbolType.Text);
+                    TagContent(); // Parse the tag, don't care about the content
+                    Optional(HtmlSymbolType.ForwardSlash);
+                    Optional(HtmlSymbolType.CloseAngle);
+
+                    if (scriptTag)
                     {
-                        // Parsing an end tag
-                        // This section can accept things like: '</p  >' or '</p>' etc.
-                        Optional(HtmlSymbolType.ForwardSlash);
-                        // Whitespace here is invalid (according to the spec)
-                        Optional(HtmlSymbolType.Text);
-                        AcceptAll(HtmlSymbolType.WhiteSpace);
-                        Optional(HtmlSymbolType.CloseAngle);
-                    }
+                        Output(SpanKind.Markup);
+                        tagBlock.Dispose();
 
-                    Output(SpanKind.Markup);
+                        SkipToEndScriptAndParseCode();
+                        return;
+                    }
+                }
+                else
+                {
+                    // Parsing an end tag
+                    // This section can accept things like: '</p  >' or '</p>' etc.
+                    Optional(HtmlSymbolType.ForwardSlash);
 
-                    // End tag block
-                    tagBlock.Dispose();
+                    // Whitespace here is invalid (according to the spec)
+                    OptionalBangEscape();
+                    Optional(HtmlSymbolType.Text);
+                    AcceptAll(HtmlSymbolType.WhiteSpace);
+                    Optional(HtmlSymbolType.CloseAngle);
                 }
+
+                Output(SpanKind.Markup);
+
+                // End tag block
+                tagBlock.Dispose();
+            }
+        }
+
+        private bool AtBangEscape(int lookahead = 0)
+        {
+            var potentialBang = lookahead == 0 ? CurrentSymbol : Lookahead(lookahead);
```

Comment:
`Lookahead` throws when provided with a `0` value. Only reason why it's ok here is because we'll conditionally lookahead.

The name lookahead doesn't really make much sense when you're not really looking ahead lol.